File: android/src/main/java/com/amplitude/experiment/Experiment.java
package com.amplitude.experiment;

import android.app.Application;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.ThreadFactory;

import okhttp3.OkHttpClient;

/**
 * Factory class for Experiment. Manages a pool of singleton instances of {@link ExperimentClient} identified by
 * name. {@link ExperimentClient} names are normalized to be lowercase. The name "$default_instance" is
 * reserved for the default {@link ExperimentClient} instance, and is provided for convenience.
 * <p>
 * All {@link ExperimentClient} share the same executor service and http client.
 */
public class Experiment {

    public static final String TAG = "Experiment";

    private static final ThreadFactory DAEMON_THREAD_FACTORY = new ThreadFactory() {
        public Thread newThread(Runnable r) {
            Thread t = Executors.defaultThreadFactory().newThread(r);
            t.setDaemon(true);
            return t;
        }
    };
    private static final ScheduledThreadPoolExecutor EXECUTOR_SERVICE =
            new ScheduledThreadPoolExecutor(0
            , DAEMON_THREAD_FACTORY);
    private static final OkHttpClient HTTP_CLIENT = new OkHttpClient();
    private static final Map<String, DefaultExperimentClient> INSTANCES = new ConcurrentHashMap<String,
            DefaultExperimentClient>();

    // Public API

    /**
     * Returns the default {@link ExperimentClient} instance.
     */
    @Nullable
    public static ExperimentClient getInstance() {
        return getInstance(ExperimentConfig.Defaults.INSTANCE_NAME);
    }

    /**
     * Returns the {@link ExperimentClient} instance associated with the provided name. If no Client
     * was initialized with the given name, returns null.
     */
    @Nullable
    public static ExperimentClient getInstance(@Nullable String name) {
        String normalizedName = ExperimentConfig.normalizeInstanceName(name);
        return INSTANCES.get(normalizedName);
    }

    /**
     * Returns the default {@link ExperimentClient} instance associated with the provided name.
     * @throws IllegalStateException If the default client has not been initialized.
     */
    @NotNull
    public static ExperimentClient getInstanceOrThrow() {
        ExperimentClient c = getInstance(ExperimentConfig.Defaults.INSTANCE_NAME);
        if (c == null)
            throw new IllegalStateException("The default ExperimentClient must be initialized before accessing the default instance");
        return getInstance(ExperimentConfig.Defaults.INSTANCE_NAME);
    }

    /**
     * Returns the {@link ExperimentClient} instance.
     * @throws IllegalStateException If the default client has not been initialized.
     */
    @NotNull
    public static ExperimentClient getInstanceOrThrow(@Nullable String name) {
        ExperimentClient c = getInstance(ExperimentConfig.Defaults.INSTANCE_NAME);
        if (c == null)
            throw new IllegalStateException("The \""+name+"\" ExperimentClient must be initialized before accessing the default instance");
        return getInstance(name);
    }

    /**
     * Initializes a {@link ExperimentClient} with the provided api key and {@link ExperimentConfig}.
     * If a {@link ExperimentClient} already exists with the instanceName set by the {@link ExperimentConfig},
     * returns that instance instead.
     *
     * @param application The Android Application context
     * @param apiKey  The API key. This can be found in the Experiment settings and should not be null or empty.
     * @param config see {@link ExperimentConfig} for configuration options
     */
    @NotNull
    public static synchronized ExperimentClient init(
            @NotNull Application application,
            @NotNull String apiKey,
            @NotNull ExperimentConfig config
    ) {
        String instanceName = config.getInstanceName();
        DefaultExperimentClient client = INSTANCES.get(instanceName);
        if (client == null) {
            client = new DefaultExperimentClient(apiKey, config,
                    new SharedPrefsStorage(application, instanceName), HTTP_CLIENT, EXECUTOR_SERVICE);
            INSTANCES.put(instanceName, client);
        }
        return client;
    }

    /**
     * Shuts down all {@link ExperimentClient} instances.
     */
    public static void shutdown() {
        HTTP_CLIENT.dispatcher().executorService().shutdown();
    }
}

File: sdk/src/main/java/com/amplitude/experiment/Experiment.kt
package com.amplitude.experiment

import android.app.Application
import com.amplitude.analytics.connector.AnalyticsConnector
import com.amplitude.experiment.storage.SharedPrefsStorage
import com.amplitude.experiment.util.AndroidLogger
import com.amplitude.experiment.util.Logger
import okhttp3.OkHttpClient
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.ThreadFactory

object Experiment {

    private val daemonThreadFactory =
        ThreadFactory { r ->
            Executors.defaultThreadFactory().newThread(r).apply {
                isDaemon = true
            }
        }
    internal val executorService = ScheduledThreadPoolExecutor(2, daemonThreadFactory)

    private val httpClient = OkHttpClient()
    private val instances = mutableMapOf<String, ExperimentClient>()

    /**
     * Initializes a singleton [ExperimentClient] identified by the configured
     * instance name.
     *
     * @param application The Android Application context
     * @param apiKey  The API key. This can be found in the Experiment settings
     * and should not be null or empty.
     * @param config see [ExperimentConfig] for configuration options
     */
    @JvmStatic
    fun initialize(
        application: Application,
        apiKey: String,
        config: ExperimentConfig
    ): ExperimentClient = synchronized(instances) {
        val instanceName = config.instanceName
        val instanceKey = "$instanceName.$apiKey"
        return when (val instance = instances[instanceKey]) {
            null -> {
                Logger.implementation = AndroidLogger(config.debug)
                var mergedConfig = config
                if (config.userProvider == null) {
                    mergedConfig = config.copyToBuilder()
                        .userProvider(DefaultUserProvider(application))
                        .build()
                }
                val newInstance = DefaultExperimentClient(
                    apiKey,
                    mergedConfig,
                    httpClient,
                    SharedPrefsStorage(application),
                    executorService,
                )
                instances[instanceKey] = newInstance
                newInstance
            }
            else -> instance
        }
    }

    /**
     * Initialize a singleton [ExperimentClient] which automatically
     * integrates with the installed and initialized instance of the amplitude
     * analytics SDK.
     *
     * You must be using Amplitude-Android SDK version 2.36.0+ or
     * Amplitude-Kotlin 1.5.0+ for this integration to work.
     *
     * @param application The Android Application context
     * @param apiKey  The API key. This can be found in the Experiment settings
     * and should not be null or empty.
     * @param config see [ExperimentConfig] for configuration options
     */
    @JvmStatic
    fun initializeWithAmplitudeAnalytics(
        application: Application,
        apiKey: String,
        config: ExperimentConfig
    ): ExperimentClient = synchronized(instances) {
        val instanceName = config.instanceName
        val instanceKey = "$instanceName.$apiKey"
        val connector = AnalyticsConnector.getInstance(instanceName)
        val instance = when (val instance = instances[instanceKey]) {
            null -> {
                Logger.implementation = AndroidLogger(config.debug)
                val configBuilder = config.copyToBuilder()
                if (config.userProvider == null) {
                    configBuilder.userProvider(
                        ConnectorUserProvider(application, connector.identityStore)
                    )
                }
                if (config.exposureTrackingProvider == null) {
                    configBuilder.exposureTrackingProvider(
                        ConnectorExposureTrackingProvider(connector.eventBridge)
                    )
                }
                val newInstance = DefaultExperimentClient(
                    apiKey,
                    configBuilder.build(),
                    httpClient,
                    SharedPrefsStorage(application),
                    executorService,
                )
                instances[instanceKey] = newInstance
                if (config.automaticFetchOnAmplitudeIdentityChange) {
                    connector.identityStore.addIdentityListener {
                        newInstance.fetch()
                    }
                }
                newInstance
            }
            else -> instance
        }
        return instance
    }
}
